File: js/35.js
const searchInsert = (nums, target) => {
  let low = 0;
  let high = nums.length - 1;
  let count;
  while (low <= high) {
    let middle = low + ((high - low) % 2);
    if (target === nums[middle]) {
      return middle;
    } else if (target > nums[middle]) {
      low = middle + 1;
      count = low;
    } else if (target < nums[middle]) {
      high = middle - 1;
      if (nums[high] < target) {
        count = high + 1;
      } else {
        count = high;
      }
    }
  }
  return count > 0 ? count : 0;
};

console.log(searchInsert([1, 3, 5, 6], 5));
console.log(searchInsert([1, 3, 5, 6], 7));
console.log(searchInsert([1, 3, 5, 6], 2));
console.log(searchInsert([1, 3, 5, 6], 0));
console.log(searchInsert([1, 3], 2));

File: py/35.py
from typing import List


def search_insert(nums: List[int], target: int) -> int:
    low = 0
    high = len(nums) - 1
    count = 0
    while low <= high:
        middle = low + (high - low) // 2
        if target == nums[middle]:
            return middle
        elif target > nums[middle]:
            low = middle + 1
            count = low
        elif target < nums[middle]:
            high = middle - 1
            if nums[high] < target:
                count = high + 1
            else:
                count = high
    return count if count > 0 else 0


print(search_insert([1, 3, 5, 6], 5))
print(search_insert([1, 3, 5, 6], 7))
print(search_insert([1, 3, 5, 6], 2))
print(search_insert([1, 3, 5, 6], 0))
print(search_insert([1, 3], 2))
